trl/trainer/srpo_trainer.py

```diff
@@ -0,0 +1,1756 @@
+        """
+        if eval_dataset is None and self.eval_dataset is None:
+            raise ValueError("Trainer: evaluation requires an eval_dataset.")
+        eval_dataset = eval_dataset if eval_dataset is not None else self.eval_dataset
+
+        if self.precompute_ref_log_probs and not self._precomputed_eval_ref_log_probs:
+            dataloader_params = {
+                "batch_size": self.args.per_device_eval_batch_size,
+                "collate_fn": self.data_collator,
+                "num_workers": self.args.dataloader_num_workers,
+                "pin_memory": self.args.dataloader_pin_memory,
+                "shuffle": False,
+            }
+
+            # prepare dataloader
+            data_loader = self.accelerator.prepare(
+                DataLoader(eval_dataset, **dataloader_params)
+            )
+
+            reference_chosen_logps = []
+            reference_rejected_logps = []
+            for padded_batch in tqdm(
+                iterable=data_loader, desc="Eval dataset reference log probs"
+            ):
+                reference_chosen_logp, reference_rejected_logp = (
+                    self.compute_reference_log_probs(padded_batch)
+                )
+                reference_chosen_logp, reference_rejected_logp = (
+                    self.accelerator.gather_for_metrics(
+                        (reference_chosen_logp, reference_rejected_logp)
+                    )
+                )
+                reference_chosen_logps.append(reference_chosen_logp.cpu())
+                reference_rejected_logps.append(reference_rejected_logp.cpu())
+
+            all_reference_chosen_logps = (
+                torch.cat(reference_chosen_logps).float().numpy()
+            )
+            all_reference_rejected_logps = (
+                torch.cat(reference_rejected_logps).float().numpy()
+            )
+
+            eval_dataset = eval_dataset.add_column(
+                name="reference_chosen_logps", column=all_reference_chosen_logps
+            )
+            eval_dataset = eval_dataset.add_column(
+                name="reference_rejected_logps", column=all_reference_rejected_logps
+            )
+
+            # Save calculated reference_chosen_logps and reference_rejected_logps to the eval_dataset for subsequent runs
+            if self.eval_dataset is not None:
+                self.eval_dataset = eval_dataset
+            self._precomputed_eval_ref_log_probs = True
+
+        return super().get_eval_dataloader(eval_dataset=eval_dataset)
+
+    def build_tokenized_answer(self, prompt, answer, example=None):
+        """
+        Llama tokenizer does satisfy `enc(a + b) = enc(a) + enc(b)`.
+        It does ensure `enc(a + b) = enc(a) + enc(a + b)[len(enc(a)):]`.
+        Reference:
+            https://github.com/EleutherAI/lm-evaluation-harness/pull/531#issuecomment-1595586257
+        """
+
+        if example:
+            full = self.tokenizer.apply_chat_template(
```

Comment:
I used the chat template within the `build_tokenized_answer` field and assumed that we accepted `example` and `answer`. Open to discussion on better ways to handle this.